rust/core-lib/src/movement.rs

```diff
@@ -25,43 +29,158 @@ pub enum Movement {
+
 fn region_movement(m: Movement, r: &SelRegion, view: &View, text: &Rope, modify: bool)
     -> (usize, Option<HorizPos>)
 {
     match m {
         Movement::Left => {
             if r.is_caret() || modify {
                 if let Some(offset) = text.prev_grapheme_offset(r.end) {
-                    calc_horiz(view, text, offset)
+                    (offset, None)
                 } else {
-                    (0, None)
+                    (0, r.horiz)
                 }
             } else {
-                calc_horiz(view, text, r.min())
+                (r.min(), None)
             }
         }
         Movement::Right => {
             if r.is_caret() || modify {
                 if let Some(offset) = text.next_grapheme_offset(r.end) {
-                    calc_horiz(view, text, offset)
+                    (offset, None)
                 } else {
-                    (r.end, None)
+                    (r.end, r.horiz)
                 }
             } else {
-                calc_horiz(view, text, r.max())
+                (r.max(), None)
+            }
+        }
+        Movement::LeftWord => {
+            let mut word_cursor = WordCursor::new(text, r.end);
+            let offset = word_cursor.prev_boundary().unwrap_or(0);
+            (offset, None)
+        }
+        Movement::RightWord => {
+            let mut word_cursor = WordCursor::new(text, r.end);
+            let offset = word_cursor.next_boundary().unwrap_or_else(|| text.len());
+            (offset, None)
+        }
+        Movement::LeftOfLine => {
+            let line = view.line_of_offset(text, r.end);
+            let offset = view.offset_of_line(text, line);
+            (offset, None)
+        }
+        Movement::RightOfLine => {
+            let line = view.line_of_offset(text, r.end);
+            let mut offset = text.len();
+
+            // calculate end of line
+            let next_line_offset = view.offset_of_line(text, line + 1);
+            if line < view.line_of_offset(text, offset) {
+                if let Some(prev) = text.prev_grapheme_offset(next_line_offset) {
+                    offset = prev;
+                }
+            }
+            (offset, None)
+        }
+        Movement::Up => vertical_motion(r, view, text, -1, modify),
+        Movement::Down => vertical_motion(r, view, text, 1, modify),
+        Movement::StartOfParagraph => {
+            // Note: TextEdit would start at modify ? r.end : r.min()
+            let mut cursor = Cursor::new(&text, r.end);
+            let offset = cursor.prev::<LinesMetric>().unwrap_or(0);
+            (offset, None)
+        }
+        Movement::EndOfParagraph => {
+            // Note: TextEdit would start at modify ? r.end : r.max()
+            let mut offset = r.end;
+            let mut cursor = Cursor::new(&text, offset);
+            if let Some(next_para_offset) = cursor.next::<LinesMetric>() {
+                if cursor.is_boundary::<LinesMetric>() {
```

Comment:
to clarify: if `cursor.is_boundary() == false` in this situation we're at the end of the file?